source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -316,6 +305,117 @@ void RedisCluster::RedisDiscoverySession::startResolveRedis() {
+  } else if (empty_response) {
+    parent_.info_->stats().update_empty_.inc();
+  }
+}
+
+/**
+ * Resolve the primary cluster entry hostname in each slot.
+ * If the primary is successfully resolved, we proceed to resolve replicas.
+ * We use the count of hostnames that require resolution to decide when the resolution process is
+ * completed, and then call the post-resolution hooks.
+ *
+ * @param slots the list of slots which may need DNS resolution
+ * @param address_resolution_required_cnt the number of hostnames that need DNS resolution
+ */
+void RedisCluster::RedisDiscoverySession::resolveClusterHostnames(
+    ClusterSlotsSharedPtr&& slots,
+    std::shared_ptr<std::uint64_t> hostname_resolution_required_cnt) {
+  for (uint64_t slot_idx = 0; slot_idx < slots->size(); slot_idx++) {
+    auto& slot = (*slots)[slot_idx];
+    if (slot.primary() == nullptr) {
+      ENVOY_LOG(debug,
+                "starting async DNS resolution for primary slot address {} at index location {}",
+                slot.primary_hostname_, slot_idx);
+      parent_.dns_resolver_->resolve(
+          slot.primary_hostname_, parent_.dns_lookup_family_,
+          [this, slot_idx, slots,
+           &hostname_resolution_required_cnt](Network::DnsResolver::ResolutionStatus status,
+                                              std::list<Network::DnsResponse>&& response) -> void {
+            auto& slot = (*slots)[slot_idx];
+            ENVOY_LOG(debug, "async DNS resolution complete for {}", slot.primary_hostname_);
+            updateDnsStats(status, response.empty());
+            // If DNS resolution for a primary fails, we stop resolution for remaining, and reset
+            // the timer.
+            if (status != Network::DnsResolver::ResolutionStatus::Success) {
+              ENVOY_LOG(error, "Unable to resolve cluster slot primary hostname {}",
+                        slot.primary_hostname_);
+              resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+              return;
+            }
+            // Primary slot address resolved
+            slot.setPrimary(Network::Utility::getAddressWithPort(
+                *response.front().addrInfo().address_, slot.primary_port_));
+            (*hostname_resolution_required_cnt)--;
+            // Continue on to resolve replicas
+            resolveReplicas(slots, slot_idx, hostname_resolution_required_cnt);
+          });
+    } else {
+      resolveReplicas(slots, slot_idx, hostname_resolution_required_cnt);
+    }
+  }
+}
+
+/**
+ * Resolve the replicas in a cluster entry. If there are no replicas, simply return.
+ * If all the hostnames have been resolved, call post-resolution methods.
+ *
+ * @param slots the list of slots which may need DNS resolution
+ * @param index the specific index into `slots` whose replicas need to be resolved
+ * @param address_resolution_required_cnt the number of address that need to be resolved
+ */
+void RedisCluster::RedisDiscoverySession::resolveReplicas(
```

Comment:
For my understanding, why do we resolve replicas after resolving primaries? Can't it all be done in parallel? Also, from my read of the code, not resolving primaries is a fatal error, but not not resolving replicas is not? can you clarify that in comments and docs?

Reply:
> For my understanding, why do we resolve replicas after resolving primaries? Can't it all be done in parallel?

My motivations for resolving replicas after primary is so that we can skip resolving replicas if the primary resolution fails. I guess we can do it all in parallel. The current flow was just easier to reason about in my head. I wasn't sure if the potential gains are worth it.

>Also, from my read of the code, not resolving primaries is a fatal error, but not not resolving replicas is not? can you clarify that in comments and docs?

You are right. Failure to resolve any one primary will stop the resolution process, and reset the timer to find the primaries. I'll add that to the comments here. For documentation, is it the same as https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_protocols/redis.html ?

Reply:
Yes please add something about the new feature to that page as well as add a release note for the overall feature in current.rst. Thank you.

Reply:
(I think what you have now about the 2 stage resolve is fine, though it might happen a bit faster if all is done in parallel. Up to you.)